### PR TITLE
Feature 1 - Render a game state

### PR DESCRIPTION
**Pull request by atilbian**
- [x] Draw a game from a GameState object
- [x] Print the rendered game in the console
- [x] Test it
- [x] Make the expected game more readable in UT

### CODE REVIEW

File: src/render.spec.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    board: [
+      [{ value: CellValue.Empty }, { value: CellValue.Black }],
+      [{ value: CellValue.Empty }, { value: CellValue.Black }],
+    ],
+  };
+}
+
+describe("Get 2-rows board with all cells empty must return it", () => {
+  it("should pass", () => {
+    expect(getRenderedGameState(get2RowsGameStateAllCellsEmpty())).toBe(
```
Comment (fzaninotto) [inline src/render.spec.ts]: not readable: I need to move up and down to look at the test and the input.

File: src/render.spec.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      "⬡⬡\n ⬡⬡"
```
Comment (fzaninotto) [inline src/render.spec.ts]: Not readable. Please use backticks to that the expected outputs looks like an actual board

File: src/render.spec.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+function get2RowsGameStateMixedCellsValues(): GameState {
+  return {
+    board: [
+      [{ value: CellValue.Empty }, { value: CellValue.Black }],
+      [{ value: CellValue.Empty }, { value: CellValue.Black }],
+    ],
+  };
+}
+
+describe("Get 2-rows board with all cells empty must return it", () => {
```
Comment (fzaninotto) [inline src/render.spec.ts]: We usually wite one "describe" per main function, and many "it" inside for each case

File: src/render.spec.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    );
+  });
+});
+
+describe("Get 2-rows board with mixed cells values must return it", () => {
+  it("should pass", () => {
+    expect(getRenderedGameState(get2RowsGameStateMixedCellsValues())).toBe(
+      "⬡⬢\n ⬡⬢"
+    );
+  });
```
Comment (fzaninotto) [inline src/render.spec.ts:105]: you're probbaly missing a test with a bigger board.

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: src/render.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    const initialSpace = spaceCharacter.repeat(index);
+    let renderedRow = initialSpace;
+    row.forEach((cell) => {
+      const renderedCell = cell.value == CellValue.Black ? "⬢" : "⬡";
+      renderedRow += renderedCell;
+    });
+    if (index > 0) {
+      result += lineSeparatorCharacter;
+    }
+    result += renderedRow;
```
Comment (slax57) [inline src/render.ts]:
```suggestion
    result += renderedRow;
    result += lineSeparatorCharacter;
```

File: src/render.spec.ts (hunk 2 above)
Comment (atilbian) [reply src/render.spec.ts]: I'm not sure to understand how to do it with backticks. Could we check that tomorrow?

File: src/gamestate.ts [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
   board: Array<Array<Cell>>;
 }
 
-interface Cell {
+export interface Cell {
   value: CellValue;
 }
 
-enum CellValue {
+export enum CellValue {
```
Comment (mchaffotte) [inline src/gamestate.ts:9]: use union instead of enum

File: src/render.spec.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+describe("Draw a boad from a game state object", () => {
+  it("Draw 2-rows board with all cells empty", () => {
+    const twoRowsAllCellsEmpty: GameState = {
+      board: [
+        [{ value: CellValue.Empty }, { value: CellValue.Empty }],
+        [{ value: CellValue.Empty }, { value: CellValue.Empty }],
+      ],
+    };
+    expect(getRenderedGameState(twoRowsAllCellsEmpty)).toBe("⬡⬡\n ⬡⬡\n");
```
Comment (mchaffotte) [inline src/render.spec.ts]: Use the Given/When/The pattern in order to ease the test readability

File: src/render.spec.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import expect from "expect";
+import { getRenderedGameState } from "./render";
+import { GameState, CellValue } from "./gamestate";
+
+describe("Draw a boad from a game state object", () => {
+  it("Draw 2-rows board with all cells empty", () => {
```
Comment (mchaffotte) [inline src/render.spec.ts]: it("should draw .....", ....

File: src/render.spec.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        [{ value: CellValue.Empty }, { value: CellValue.Black }],
+        [{ value: CellValue.Empty }, { value: CellValue.Black }],
+        [{ value: CellValue.Empty }, { value: CellValue.Black }],
+      ],
+    };
+    expect(getRenderedGameState(fiveRowsWithMixedCellValues)).toBe(
+      "⬡⬢\n ⬡⬢\n  ⬡⬢\n   ⬡⬢\n    ⬡⬢\n"
+    );
+  });
+});
```
Comment (mchaffotte) [inline src/render.spec.ts:106]: You should think about tests using limit cases / exceptions (for example: with an empty array)

File: src/render.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import { GameState, CellValue } from "./gamestate";
+
+export function displayGameState(gameState: GameState) {
+  const renderedGameState = getRenderedGameState(gameState);
+  console.log(renderedGameState);
+}
+
+export function getRenderedGameState(gameState: GameState): string {
+  const spaceCharacter = " ";
+  const lineSeparatorCharacter = "\n";
```
Comment (mchaffotte) [inline src/render.ts]: extract constants outside of the function (-> each time the function is called these variables are built again)

File: src/render.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+  let result = "";
+  gameState.board.forEach((row, index) => {
+    const initialSpace = spaceCharacter.repeat(index);
+    let renderedRow = initialSpace;
+    row.forEach((cell) => {
+      const renderedCell = cell.value == CellValue.Black ? "⬢" : "⬡";
```
Comment (mchaffotte) [inline src/render.ts]: extract your pawns in contants

Comment (mchaffotte) [inline src/render.ts]: It should be better to extract the inner loop for a better readability

**Review by mchaffotte — CHANGES_REQUESTED** [review]

File: src/render.ts (hunk 11 above)
Comment (slax57) [reply src/render.ts]: names should be uppercase?

File: src/gamestate.ts (hunk 6 above)
Comment (slax57) [reply src/gamestate.ts:9]: will be resolved with https://github.com/marmelab/Hex/pull/3#discussion_r801038301

File: src/render.spec.ts (hunk 7 above)
Comment (atilbian) [reply src/render.spec.ts]: Am i using it correctly ?

```
const emptyBoard = {
      board: [],
};

const renderedGameState = getRenderedGameState(emptyBoard);

expect(renderedGameState).toBe("");
```

File: src/render.ts (hunk 11 above)
Comment (mchaffotte) [reply src/render.ts]: Yes it must be ;)

**Pull request merged**